### PR TITLE
Enabling Multicluster dashboard when ACM observability is enabled - WIP

### PR DESCRIPTION
**Pull request by GowthamShanmugam**
on top of: https://github.com/red-hat-storage/odf-console/pull/156

### CODE REVIEW

File: packages/mco/components/mco-dashboard/queries.ts [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+    [StorageDashboard.TOTAL_CAP_FILE_BLOCK]:
+      '(label_replace(odf_system_map{target_namespace="openshift-storage"} , "managedBy", "$1", "target_name", "(.*)"))  * on (namespace, managedBy) group_right(storage_system) odf_system_raw_capacity_total_bytes',
+    [StorageDashboard.USED_CAPACITY_FILE_BLOCK]:
+      '(label_replace(odf_system_map{target_namespace="openshift-storage"} , "managedBy", "$1", "target_name", "(.*)"))  * on (namespace, managedBy) group_right(storage_system) odf_system_raw_capacity_used_bytes',
+  };
+  
+  export const STATUS_QUERIES = {
+    [StorageDashboard.HEALTH]: '(label_replace(odf_system_map{target_namespace="openshift-storage"} , "managedBy", "$1", "target_name", "(.*)"))  * on (namespace, managedBy) group_right(storage_system) odf_system_health_status'
+  };
+
```
Comment (bipuladh) [inline packages/mco/components/mco-dashboard/queries.ts:17]: new line missing

File: packages/mco/components/mco-dashboard/status-card/status-card.tsx [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+    case '0':
+      return HealthState.OK;
+    case '1':
+      return HealthState.WARNING;
+    case '2':
+      return HealthState.ERROR;
+    default:
+      return HealthState.LOADING;
+  }
+};
```
Comment (bipuladh) [inline packages/mco/components/mco-dashboard/status-card/status-card.tsx:33]: Refactor this from ODF package and put it in shared.

File: packages/mco/components/mco-dashboard/status-card/status-card.tsx [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+
+  const [worstHealth, setWorstHealth] = React.useState<string>('0');
+  const [healthData, healthError, healthLoading] =
+    useURLPoll<PrometheusResponse>(`${ACM_ENDPOINT}=${STATUS_QUERIES[StorageDashboard.HEALTH]}`);
+
+  const parsedHealthData = React.useMemo(() => {
+      let health = '0';
+      let data = !healthError && !healthLoading
+        ? healthData?.data?.result.reduce((acc, statusMetric) => {
+            health = health < statusMetric?.value[1] ? statusMetric?.value[1] : health;
```
Comment (bipuladh) [inline packages/mco/components/mco-dashboard/status-card/status-card.tsx:46]: Ideally reduce should not have side effects to variables outside it's scope. So I don't think reduce or map should be used in this case.

Comment (bipuladh) [inline packages/mco/components/mco-dashboard/status-card/status-card.tsx:46]: health is a string in the outer scope, I am not sure even if intrinsic conversion happens it's a right way to do the comparisons. Use  explicit conversions is required?

File: packages/mco/components/mco-dashboard/status-card/status-card.tsx [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+          <Flex
+            justifyContent={{ default: 'justifyContentSpaceBetween' }}
+            className="odf-StatusCard__items--width"
+          >
+            <FlexItem>
+              <HealthItem
+                title={t('Openshift Data Foundation')}
+                state={HealthState.OK}
+              >
+                {/*<StorageSystemPopup systemHealthMap={healthySystems} />*/}
```
Comment (bipuladh) [inline packages/mco/components/mco-dashboard/status-card/status-card.tsx:74]: Why?

File: packages/mco/components/mco-dashboard/status-card/storage-system-popup.tsx [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+
+const getSystemHealthCount = (systemHealthMap: SystemHealthMap[]) => {
+    let critical, warning, normal;
+    critical = warning = normal = 0;
+    systemHealthMap.forEach((item) => {
+        if (item?.rawHealthData === '0') normal++;
+        else if (item?.rawHealthData === '1') warning++;
+        else if (item?.rawHealthData === '2') critical++;
+    });
+    return {critical, warning, normal};
```
Comment (bipuladh) [inline packages/mco/components/mco-dashboard/status-card/storage-system-popup.tsx:47]: Use reduce here.

File: packages/mco/components/mco-dashboard/status-card/storage-system-popup.tsx [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+  systemHealthMap: SystemHealthMap[];
+};
+
+type SystemHealthCountProps = {
+    value: string;
+    health: HealthState;
+    count: number;
+}
+
+const healthStateToIcon = {
```
Comment (bipuladh) [inline packages/mco/components/mco-dashboard/status-card/storage-system-popup.tsx:27]: How about using https://github.com/red-hat-storage/odf-console/blob/master/packages/shared/dashboards/status-card/states.tsx#L42 ?

File: packages/odf/features.ts [hunk 7]
```diff
@@ -94,3 +101,49 @@ export const setOCSFlags = async (setFlag: SetFeatureFlag) => {
...
+    k8sList({model: mcokind, queryParams: {}})
+      .then((data: ACMMultiClusterObservability[]) => {
+        const isEnabled = data?.[0].status?.conditions?.some((condition) => condition.status == "True"  && condition.type == "Ready");
+        if (isEnabled) {
+          k8sGet({model: configMapkind, name: ACM_WHITELISTING_METRICS_CONFIG_MAP, ns: ACM_OBSERVABILITY_NS}).then((config: ConfigMapKind) => {
+            if (config?.metadata?.labels?.[ODF_METRICS_LABEL] === "true"){
+              setFeatureFlag(ACM_OBSERVABILITY_FLAG, true);
+            };
+          }).catch((error) => {
+
```
Comment (bipuladh) [inline packages/odf/features.ts:122]: console.error or do something here.

File: packages/odf/features.ts [hunk 8]
```diff
@@ -94,3 +101,49 @@ export const setOCSFlags = async (setFlag: SetFeatureFlag) => {
   ocsDetector();
   ocsIntervalId = setInterval(ocsDetector, 15 * SECOND);
 };
+
+
+export const detectACMObservability = async (setFeatureFlag: SetFeatureFlag) => {
+  let id = null;
+  let isInitial = true;
+  let mcokind: K8sModel = {...AcmMultiClusterObservabilityModel}
```
Comment (bipuladh) [inline packages/odf/features.ts:109]:
```suggestion
  let mcokind: K8sModel = AcmMultiClusterObservabilityModel
```
Wouldn't this work?
Same for configMapKind

**Review by bipuladh — CHANGES_REQUESTED** [review]

File: packages/odf/features.ts [hunk 9]
```diff
@@ -1,6 +1,11 @@
 import { OCSStorageClusterModel } from '@odf/core/models';
-import { StorageClusterKind } from '@odf/shared/types';
-import { SetFeatureFlag, k8sList } from "@openshift-console/dynamic-plugin-sdk";
+import { ACM_OBSERVABILITY_NS, ACM_WHITELISTING_METRICS_CONFIG_MAP, ODF_METRICS_LABEL  } from '@odf/mco/constants';
+import { AcmMultiClusterObservabilityModel } from '@odf/mco/models';
+import { ACMMultiClusterObservability } from '@odf/mco/types';
```
Comment (SanjalKatiyar) [inline packages/odf/features.ts:4]: https://github.com/red-hat-storage/odf-console/blob/7150378e539894dcb672ef7079487faa1beb7258/packages/odf/package.json#L6
now won't `/odf` be having a dependency on `/mco` as well ??

File: packages/mco/components/mco-dashboard/status-card/status-card.tsx (hunk 2 above)
Comment (SanjalKatiyar) [reply packages/mco/components/mco-dashboard/status-card/status-card.tsx:33]: ack

File: packages/mco/components/mco-dashboard/status-card/status-card.tsx (hunk 3 above)
Comment (SanjalKatiyar) [reply packages/mco/components/mco-dashboard/status-card/status-card.tsx:46]: there is no type conversion here (implicit or explicit)... I am doing string comparison only...
`health` & `statusMetric?.value[1]` both are strings which can be `0 or 1 or 2` that's all, so even if we do comparison between these strings result will be same like doing comparison between numbers. So, I don't think there is need to do type conversion.

File: packages/mco/components/mco-dashboard/queries.ts (hunk 1 above)
Comment (GowthamShanmugam) [reply packages/mco/components/mco-dashboard/queries.ts:17]: @SanjalKatiyar

File: packages/mco/components/mco-dashboard/status-card/status-card.tsx (hunk 4 above)
Comment (SanjalKatiyar) [reply packages/mco/components/mco-dashboard/status-card/status-card.tsx:74]: It is WIP, we had a discussion regarding this yesterday only... will update this...

File: packages/mco/components/mco-dashboard/status-card/storage-system-popup.tsx (hunk 6 above)
Comment (SanjalKatiyar) [reply packages/mco/components/mco-dashboard/status-card/storage-system-popup.tsx:27]: ack...

File: packages/odf/features.ts (hunk 9 above)
Comment (GowthamShanmugam) [reply packages/odf/features.ts:4]: yes, mco is seperate packages now, which has only multicluster dashbaord related modals, i.e ACM and ramen models

File: packages/odf/features.ts (hunk 7 above)
Comment (GowthamShanmugam) [reply packages/odf/features.ts:122]: ack

File: packages/odf/features.ts (hunk 9 above)
Comment (GowthamShanmugam) [reply packages/odf/features.ts:4]: mmmm. it makesense. i will move this to shared

**Comment by GowthamShanmugam** [comment]
@SanjalKatiyar @bipuladh changes done on enabling/disabling dashboard part.

File: packages/shared/models/acm.ts [hunk 10]
```diff
@@ -0,0 +1,15 @@
+import { K8sModel } from '@openshift-console/dynamic-plugin-sdk/lib/api/common-types';
```
Comment (GowthamShanmugam) [inline packages/shared/models/acm.ts:1]: moved to shared because, it might be required by MCO package also in future

**Pull request closed without merging**